File: demo/seed_database.js
/**
 * Enterprise Healthcare Demo Database Seeder (Node.js 24 Native SQLite)
 * Populates demo/healthcare_demo.db with rich synthetic Medicare Part D data
 * across 2022, 2023, and 2024.
 */

const { DatabaseSync } = require("node:sqlite");
const path = require("node:path");
const fs = require("node:fs");

const DB_PATH = path.join(__dirname, "healthcare_demo.db");

if (fs.existsSync(DB_PATH)) {
  fs.unlinkSync(DB_PATH);
}

const db = new DatabaseSync(DB_PATH);

console.log("Creating tables in demo/healthcare_demo.db...");

db.exec(`
  DROP TABLE IF EXISTS drug_summary;
  DROP TABLE IF EXISTS prescriber_summary;
  DROP TABLE IF EXISTS state_kpi;

  CREATE TABLE drug_summary (
      gnrc_name TEXT,
      brnd_name TEXT,
      year INTEGER,
      is_generic INTEGER,
      total_claims INTEGER,
      total_cost_usd REAL,
      total_beneficiaries INTEGER,
      avg_cost_per_claim REAL,
      unique_prescribers INTEGER,
      cost_rank INTEGER
  );

  CREATE TABLE prescriber_summary (
      prscrbr_npi INTEGER,
      year INTEGER,
      prscrbr_last_org_name TEXT,
      prscrbr_first_name TEXT,
      prscrbr_state_abrvtn TEXT,
      prscrbr_type TEXT,
      prscrbr_city TEXT,
      total_claims INTEGER,
      total_cost_usd REAL,
      total_beneficiaries INTEGER,
      unique_drugs_prescribed INTEGER,
      generic_rate REAL,
      state_rank INTEGER
  );

  CREATE TABLE state_kpi (
      state_abrvtn TEXT,
      year INTEGER,
      total_claims INTEGER,
      total_cost_usd REAL,
      total_beneficiaries INTEGER,
      total_prescribers INTEGER,
      unique_drugs INTEGER,
      avg_cost_per_claim REAL,
      cost_per_beneficiary REAL,
      national_rank INTEGER,
      pain_specialty_claims INTEGER
  );
`);

const STATES = [
  "TX", "CA", "FL", "NY", "PA", "OH", "IL", "GA", "NC", "MI",
  "AZ", "WA", "TN", "MA", "IN", "MO", "MD", "WI", "CO", "MN"
];

const DRUGS = [
  { gnrc: "LISINOPRIL", brnd: "PRINIVIL", is_generic: 1, base_cost: 8.5 },
  { gnrc: "ATORVASTATIN", brnd: "LIPITOR", is_generic: 1, base_cost: 12.0 },
  { gnrc: "METFORMIN", brnd: "GLUCOPHAGE", is_generic: 1, base_cost: 6.0 },
  { gnrc: "AMLODIPINE", brnd: "NORVASC", is_generic: 1, base_cost: 9.0 },
  { gnrc: "OMEPRAZOLE", brnd: "PRILOSEC", is_generic: 1, base_cost: 10.5 },
  { gnrc: "LEVOTHYROXINE", brnd: "SYNTHROID", is_generic: 1, base_cost: 11.0 },
  { gnrc: "GABAPENTIN", brnd: "NEURONTIN", is_generic: 1, base_cost: 14.0 },
  { gnrc: "HYDROCHLOROTHIAZIDE", brnd: "MICROZIDE", is_generic: 1, base_cost: 7.5 },
  { gnrc: "SERTRALINE", brnd: "ZOLOFT", is_generic: 1, base_cost: 13.0 },
  { gnrc: "OXYCODONE", brnd: "OXYCONTIN", is_generic: 0, base_cost: 185.0 },
  { gnrc: "INSULIN GLARGINE", brnd: "LANTUS", is_generic: 0, base_cost: 320.0 },
  { gnrc: "ADALIMUMAB", brnd: "HUMIRA", is_generic: 0, base_cost: 5800.0 },
  { gnrc: "APIXABAN", brnd: "ELIQUIS", is_generic: 0, base_cost: 520.0 },
  { gnrc: "EMPAGLIFLOZIN", brnd: "JARDIANCE", is_generic: 0, base_cost: 580.0 },
  { gnrc: "SEMAGLUTIDE", brnd: "OZEMPIC", is_generic: 0, base_cost: 935.0 },
];

const SPECIALTIES = [
  "Family Medicine", "Internal Medicine", "Cardiology", "Endocrinology",
  "Pain Management", "Anesthesiology", "Psychiatry", "Nurse Practitioner"
];

const LAST_NAMES = ["Smith", "Johnson", "Williams", "Brown", "Garcia", "Miller", "Davis", "Rodriguez", "Martinez", "Wilson", "Anderson", "Taylor", "Thomas", "Moore"];
const FIRST_NAMES = ["James", "Mary", "Robert", "Patricia", "John", "Jennifer", "Michael", "Linda", "David", "Elizabeth", "Sarah", "Priya", "Wei", "Aadil"];
const CITIES = { TX: "Austin", CA: "Sacramento", FL: "Miami", NY: "Albany", PA: "Philadelphia", OH: "Columbus", IL: "Chicago", GA: "Atlanta" };

const YEARS = [2022, 2023, 2024];

// Deterministic pseudo-random helper
let seed = 42;
function random() {
  seed = (seed * 9301 + 49297) % 233280;
  return seed / 233280;
}
function randInt(min, max) {
  return Math.floor(min + random() * (max - min + 1));
}
function choice(arr) {
  return arr[Math.floor(random() * arr.length)];
}

// 1. Seed drug summary
const insertDrug = db.prepare(`
  INSERT INTO drug_summary (gnrc_name, brnd_name, year, is_generic, total_claims, total_cost_usd, total_beneficiaries, avg_cost_per_claim, unique_prescribers, cost_rank)
  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
`);

for (const year of YEARS) {
  const yearDrugs = [];
  for (const d of DRUGS) {
    const claims = d.is_generic ? randInt(15000, 400000) : randInt(2500, 50000);
    const variance = 0.9 + random() * 0.2;
    const costPerClaim = Math.round(d.base_cost * variance * 100) / 100;
    const totalCost = Math.round(claims * costPerClaim * 100) / 100;
    const benes = Math.floor(claims * (0.6 + random() * 0.3));
    const prescribers = randInt(400, 15000);

    yearDrugs.push({
      gnrc: d.gnrc,
      brnd: d.brnd,
      year,
      is_generic: d.is_generic,
      claims,
      totalCost,
      benes,
      costPerClaim,
      prescribers
    });
  }

  yearDrugs.sort((a, b) => b.totalCost - a.totalCost);
  yearDrugs.forEach((d, idx) => {
    insertDrug.run(d.gnrc, d.brnd, d.year, d.is_generic, d.claims, d.totalCost, d.benes, d.costPerClaim, d.prescribers, idx + 1);
  });
}

// 2. Seed prescribers and state KPI
const insertPrescriber = db.prepare(`
  INSERT INTO prescriber_summary (prscrbr_npi, year, prscrbr_last_org_name, prscrbr_first_name, prscrbr_state_abrvtn, prscrbr_type, prscrbr_city, total_claims, total_cost_usd, total_beneficiaries, unique_drugs_prescribed, generic_rate, state_rank)
  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
`);

const insertState = db.prepare(`
  INSERT INTO state_kpi (state_abrvtn, year, total_claims, total_cost_usd, total_beneficiaries, total_prescribers, unique_drugs, avg_cost_per_claim, cost_per_beneficiary, national_rank, pain_specialty_claims)
  VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
`);

let npiCounter = 1000000000;

for (const year of YEARS) {
  const stateAgg = {};
  for (const st of STATES) {
    stateAgg[st] = { claims: 0, cost: 0, benes: 0, prescribers: 0, painClaims: 0, rows: [] };
  }

  for (let i = 0; i < 600; i++) {
    npiCounter += randInt(1, 5);
    const state = choice(STATES);
    const specialty = choice(SPECIALTIES);
    const claims = randInt(100, 8000);
    const cost = Math.round(claims * (20 + random() * 380) * 100) / 100;
    const benes = Math.floor(claims * (0.5 + random() * 0.35));
    const uniqueDrugs = randInt(5, 25);
    const genericRate = Math.round((50 + random() * 42) * 10) / 10;
    const isPain = specialty === "Pain Management" || specialty === "Anesthesiology";

    const pRow = {
      npi: npiCounter,
      year,
      lastName: choice(LAST_NAMES),
      firstName: choice(FIRST_NAMES),
      state,
      specialty,
      city: CITIES[state] || "Metro Area",
      claims,
      cost,
      benes,
      uniqueDrugs,
      genericRate
    };

    stateAgg[state].rows.push(pRow);
    stateAgg[state].claims += claims;
    stateAgg[state].cost += cost;
    stateAgg[state].benes += benes;
    stateAgg[state].prescribers += 1;
    if (isPain) {
      stateAgg[state].painClaims += claims;
    }
  }

  // Insert prescribers with state rank
  for (const st of STATES) {
    const sObj = stateAgg[st];
    sObj.rows.sort((a, b) => b.cost - a.cost);
    sObj.rows.forEach((p, idx) => {
      insertPrescriber.run(
        p.npi, p.year, p.lastName, p.firstName, p.state, p.specialty,
        p.city, p.claims, p.cost, p.benes, p.uniqueDrugs, p.genericRate, idx + 1
      );
    });
  }

  // Insert state KPI with national rank
  const stateKeys = Object.keys(stateAgg);
  stateKeys.sort((a, b) => stateAgg[b].cost - stateAgg[a].cost);

  stateKeys.forEach((st, idx) => {
    const s = stateAgg[st];
    const avgCostPerClaim = s.claims > 0 ? Math.round((s.cost / s.claims) * 100) / 100 : 0;
    const costPerBene = s.benes > 0 ? Math.round((s.cost / s.benes) * 100) / 100 : 0;
    const uniqueDrugs = randInt(95, 145);

    insertState.run(
      st, year, s.claims, Math.round(s.cost * 100) / 100, s.benes,
      s.prescribers, uniqueDrugs, avgCostPerClaim, costPerBene, idx + 1, s.painClaims
    );
  });
}

console.log("✅ Seeded demo/healthcare_demo.db successfully!");
for (const table of ["drug_summary", "prescriber_summary", "state_kpi"]) {
  const row = db.prepare(`SELECT COUNT(*) as cnt FROM ${table}`).get();
  console.log(`   ${table.padEnd(20)} -> ${row.cnt} rows`);
}
db.close();

File: demo/seed_database.py
"""
Demo Database Seeder
Creates demo/healthcare_demo.db (SQLite) with the same logical schema as the
Snowflake GOLD_SCHEMA tables (scripts/snowflake_setup.sql), populated with
realistic synthetic data — same shape/scale-down of the real Medicare Part D
aggregates, generated deterministically (fixed random seed) so demo runs are
reproducible.

Note on auth: demo mode does NOT seed a users table here. Login already has
a working fallback for exactly this situation — api/auth/user_store.py's
get_user_by_username() tries the (Snowflake/demo-SQLite) AUTH_SCHEMA.USERS
query first, and on failure (which it always will be in demo mode, since
this seeder deliberately doesn't create that table) falls back to a
hardcoded 3-account set with real bcrypt hashes computed via the same
hash_password() function production uses. Duplicating that here would be
redundant and a second place for the demo accounts to drift out of sync.

Run: python demo/seed_database.py
"""

import sqlite3
import random
from pathlib import Path

random.seed(42)  # reproducible demo data across runs

DB_PATH = Path(__file__).parent / "healthcare_demo.db"

STATES = ["TX", "CA", "FL", "NY", "PA", "OH", "IL", "GA", "NC", "MI",
          "AZ", "WA", "TN", "MA", "IN", "MO", "MD", "WI", "CO", "MN"]

DRUGS = [
    # (generic, brand, is_generic, base_cost_per_claim)
    ("LISINOPRIL", "PRINIVIL", True, 8.5),
    ("ATORVASTATIN", "LIPITOR", True, 12.0),
    ("METFORMIN", "GLUCOPHAGE", True, 6.0),
    ("AMLODIPINE", "NORVASC", True, 9.0),
    ("OMEPRAZOLE", "PRILOSEC", True, 10.5),
    ("LEVOTHYROXINE", "SYNTHROID", True, 11.0),
    ("GABAPENTIN", "NEURONTIN", True, 14.0),
    ("HYDROCHLOROTHIAZIDE", "MICROZIDE", True, 7.5),
    ("SERTRALINE", "ZOLOFT", True, 13.0),
    ("OXYCODONE", "OXYCONTIN", False, 185.0),
    ("INSULIN GLARGINE", "LANTUS", False, 320.0),
    ("ADALIMUMAB", "HUMIRA", False, 5800.0),
    ("APIXABAN", "ELIQUIS", False, 520.0),
    ("EMPAGLIFLOZIN", "JARDIANCE", False, 580.0),
    ("SEMAGLUTIDE", "OZEMPIC", False, 935.0),
]

SPECIALTIES = [
    "Family Medicine", "Internal Medicine", "Cardiology", "Endocrinology",
    "Pain Management", "Anesthesiology", "Psychiatry", "Nurse Practitioner",
]

YEARS = [2022, 2023, 2024]


def create_schema(conn: sqlite3.Connection):
    conn.executescript("""
    DROP TABLE IF EXISTS drug_summary;
    DROP TABLE IF EXISTS prescriber_summary;
    DROP TABLE IF EXISTS state_kpi;

    CREATE TABLE drug_summary (
        gnrc_name TEXT, brnd_name TEXT, year INTEGER, is_generic INTEGER,
        total_claims INTEGER, total_cost_usd REAL, total_beneficiaries INTEGER,
        avg_cost_per_claim REAL, unique_prescribers INTEGER, cost_rank INTEGER
    );

    CREATE TABLE prescriber_summary (
        prscrbr_npi INTEGER, year INTEGER, prscrbr_last_org_name TEXT,
        prscrbr_first_name TEXT, prscrbr_state_abrvtn TEXT, prscrbr_type TEXT,
        prscrbr_city TEXT, total_claims INTEGER, total_cost_usd REAL,
        total_beneficiaries INTEGER, unique_drugs_prescribed INTEGER,
        generic_rate REAL, state_rank INTEGER
    );

    CREATE TABLE state_kpi (
        state_abrvtn TEXT, year INTEGER, total_claims INTEGER,
        total_cost_usd REAL, total_beneficiaries INTEGER, total_prescribers INTEGER,
        unique_drugs INTEGER, avg_cost_per_claim REAL, cost_per_beneficiary REAL,
        national_rank INTEGER, pain_specialty_claims INTEGER
    );
    """)


def seed_drug_summary(conn: sqlite3.Connection):
    rows = []
    for year in YEARS:
        year_costs = []
        for gnrc, brnd, is_generic, base_cost in DRUGS:
            claims = random.randint(5_000, 400_000) if is_generic else random.randint(500, 50_000)
            cost_variance = random.uniform(0.85, 1.15)
            cost_per_claim = round(base_cost * cost_variance, 2)
            total_cost = round(claims * cost_per_claim, 2)
            beneficiaries = int(claims * random.uniform(0.6, 0.9))
            prescribers = random.randint(200, 15_000)
            year_costs.append((gnrc, brnd, year, int(is_generic), claims, total_cost,
                                beneficiaries, cost_per_claim, prescribers))

        # Rank within year by total_cost_usd desc
        year_costs.sort(key=lambda r: -r[5])
        for rank, row in enumerate(year_costs, start=1):
            rows.append(row + (rank,))

    conn.executemany(
        "INSERT INTO drug_summary VALUES (?,?,?,?,?,?,?,?,?,?)", rows
    )


LAST_NAMES = ["Smith", "Johnson", "Williams", "Brown", "Garcia", "Miller", "Davis",
              "Rodriguez", "Martinez", "Wilson", "Anderson", "Taylor", "Thomas", "Moore"]
FIRST_NAMES = ["James", "Mary", "Robert", "Patricia", "John", "Jennifer", "Michael",
               "Linda", "David", "Elizabeth", "Sarah", "Priya", "Wei", "Aadil"]
CITIES = {"TX": "Austin", "CA": "Sacramento", "FL": "Miami", "NY": "Albany",
          "PA": "Philadelphia", "OH": "Columbus", "IL": "Chicago", "GA": "Atlanta"}


def seed_prescriber_and_state(conn: sqlite3.Connection):
    prescriber_rows = []
    state_agg = {}  # (state, year) -> accumulator

    npi_counter = 1_000_000_000
    for year in YEARS:
        state_prescribers = {s: [] for s in STATES}

        for _ in range(600):  # 600 prescribers per year across all states
            npi_counter += random.randint(1, 5)
            state = random.choice(STATES)
            specialty = random.choice(SPECIALTIES)
            claims = random.randint(50, 8_000)
            cost = round(claims * random.uniform(15, 400), 2)
            beneficiaries = int(claims * random.uniform(0.5, 0.85))
            unique_drugs = random.randint(3, 25)
            generic_rate = round(random.uniform(45, 92), 2)

            row = [npi_counter, year, random.choice(LAST_NAMES), random.choice(FIRST_NAMES),
                   state, specialty, CITIES.get(state, "Unknown"), claims, cost,
                   beneficiaries, unique_drugs, generic_rate]
            state_prescribers[state].append(row)

            key = (state, year)
            if key not in state_agg:
                state_agg[key] = {"claims": 0, "cost": 0.0, "benes": 0, "prescribers": 0,
                                   "drugs": set(), "pain_claims": 0}
            agg = state_agg[key]
            agg["claims"] += claims
            agg["cost"] += cost
            agg["benes"] += beneficiaries
            agg["prescribers"] += 1
            if specialty in ("Pain Management", "Anesthesiology"):
                agg["pain_claims"] += claims

        # Rank prescribers within (state, year)
        for state, rows in state_prescribers.items():
            rows.sort(key=lambda r: -r[8])  # sort by cost desc
            for rank, row in enumerate(rows, start=1):
                prescriber_rows.append(tuple(row) + (rank,))

    conn.executemany(
        "INSERT INTO prescriber_summary VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?)",
        prescriber_rows
    )

    # Build state_kpi from the aggregates collected above
    state_rows = []
    for year in YEARS:
        year_states = [(s, year) for s in STATES if (s, year) in state_agg]
        year_states.sort(key=lambda k: -state_agg[k]["cost"])
        for rank, key in enumerate(year_states, start=1):
            agg = state_agg[key]
            state, yr = key
            cost_per_bene = round(agg["cost"] / agg["benes"], 2) if agg["benes"] else 0
            avg_cost_per_claim = round(agg["cost"] / agg["claims"], 2) if agg["claims"] else 0
            unique_drugs = random.randint(80, 140)
            state_rows.append((
                state, yr, agg["claims"], round(agg["cost"], 2), agg["benes"],
                agg["prescribers"], unique_drugs, avg_cost_per_claim, cost_per_bene,
                rank, agg["pain_claims"],
            ))

    conn.executemany(
        "INSERT INTO state_kpi VALUES (?,?,?,?,?,?,?,?,?,?,?)", state_rows
    )


def main():
    if DB_PATH.exists():
        DB_PATH.unlink()

    conn = sqlite3.connect(DB_PATH)
    create_schema(conn)
    seed_drug_summary(conn)
    seed_prescriber_and_state(conn)
    conn.commit()

    # Print a summary so `python seed_database.py` gives immediate feedback
    for table in ("drug_summary", "prescriber_summary", "state_kpi"):
        count = conn.execute(f"SELECT COUNT(*) FROM {table}").fetchone()[0]
        print(f"  {table:22s} -> {count:,} rows")

    conn.close()
    print(f"\n✅ Demo database created: {DB_PATH}")
    print("   Login uses the existing hardcoded demo accounts (api/auth/user_store.py) —")
    print("   admin/Admin@123, analyst/Analyst@123, viewer/Viewer@123 — no separate seeding needed.")


if __name__ == "__main__":
    main()
